lms/services/assignment.py

```diff
@@ -0,0 +1,95 @@
+
+        Return the document URL for the assignment with the given
+        tool_consumer_instance_guid and resource_link_id, or None.
+
+        If the tool_consumer_instance_guid and resource_link_id params are
+        omitted then return the document URL of the current request's
+        assignment (from the request's tool_consumer_instance_guid and
+        resource_link_id params).
+        """
+        if tool_consumer_instance_guid is None:
+            tool_consumer_instance_guid = self._tool_consumer_instance_guid
+
+        if resource_link_id is None:
+            resource_link_id = self._resource_link_id
+
+        mic = self._get(tool_consumer_instance_guid, resource_link_id)
+
+        if mic:
+            return mic.document_url
+
+        return None
+
+    def set_document_url(
+        self, document_url, tool_consumer_instance_guid=None, resource_link_id=None
+    ):
+        """
+        Save the given document_url.
+
+        Set the document_url for the assignment that matches
+        tool_consumer_instance_guid and resource_link_id. Any existing document
+        URL for this assignment will be overwritten.
+
+        If the tool_consumer_instance_guid and resource_link_id params are
+        omitted then save document_url for the current request's assignment
+        (from the request's tool_consumer_instance_guid and resource_link_id
+        params).
+        """
+        if tool_consumer_instance_guid is None:
+            tool_consumer_instance_guid = self._tool_consumer_instance_guid
+
+        if resource_link_id is None:
+            resource_link_id = self._resource_link_id
+
+        mic = self._get(tool_consumer_instance_guid, resource_link_id)
+
+        if mic:
+            mic.document_url = document_url
+        else:
+            self._db.add(
+                ModuleItemConfiguration(
+                    document_url=document_url,
+                    resource_link_id=resource_link_id,
+                    tool_consumer_instance_guid=tool_consumer_instance_guid,
+                )
+            )
+
+        # Clear the cache (@lru_cache) on self._get because we've changed the
+        # contents of the DB. (Python's @lru_cache doesn't have a way to remove
+        # just one key from the cache, you have to clear the entire cache.)
+        self._get.cache_clear()
+
+    @lru_cache(maxsize=None)
+    def _get(self, tool_consumer_instance_guid, resource_link_id):
+        return (
+            self._db.query(ModuleItemConfiguration)
+            .filter_by(
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                resource_link_id=resource_link_id,
+            )
+            .one_or_none()
+        )
+
+
+def factory(_context, request):
+    return AssignmentService(
+        db=request.db,
+        tool_consumer_instance_guid=request.params.get("tool_consumer_instance_guid"),
+        resource_link_id=request.params.get("resource_link_id"),
```

Comment:
The `.get()`'s are necessary here because this service does get used by view predicates when serving requests that don't have these params, and if the factory crashes that causes BDD tests to fail. If the params are missing from the request then these will be `None` and `get_document_url()` will always return `None`